lib/kubernetes-deploy/kubernetes_resource.rb

```diff
@@ -474,9 +479,8 @@ def validate_spec_with_kubectl(kubectl)
       _, err, st = validate_with_dry_run_option(kubectl, "--dry-run")
```

Comment:
Two questions related to the previous PR:
- If server dry run is available, is there actually any value in running the client-side dry run as well?
- For a server-side dry run to be possible, both the resource and the server have to support it, but we seem to only look at the resource, and blindly make an API request per secret even when we could have known in advance that all of those requests will be doomed. I'm wondering if the context object we were discussing passing around (namespace, context, logger IIRC) should also contain (or be able to get) the server version for use in places like this. Was any work done on that object?

Reply:
Thanks @KnVerey for reviewing it. 
- > If server dry run is available, is there actually any value in running the client-side dry run as well?

I believe `--server-dry-run` does work including something client dry-run does. I didn't find direct answer, but from help doc:
```
$ k apply --help | grep dry-run
      --dry-run=false: If true, only print the object that would be sent, without sending it. Warning: --dry-run cannot accurately output the result of merging the local manifest and the server-side data. Use --server-dry-run to get the merged result instead.
      --server-dry-run=false: If true, request will be sent to server with dry-run flag, which means the modifications won't be persisted. This is an alpha feature and flag.
```
My local test also shows error caught during client dry-run can also be caught by server dry-run

- > For a server-side dry run to be possible ...

This is a good question. `--server-dry-run` is an alpha feature in `v1.12`, which is disable by default. It becomes to be a beta feature in `v1.13`, which enable by default then. 
We consider about checking the availability of server-dry-run by comparing server version, but we concern that if people use version after `v1.13` but cluster admins disable this feature, that'll be sad. 

But you remind me, we can do a version check before. If version is after `v1.13` and we find cluster admin disable it, we should catch feature disable error and go back to client dry-run. @KnVerey @dturn  what do you think?

Reply:
I'm not sure whether we put this into our contributor guidelines, but we had a case in the past where someone was arguing for support of an alpha resource, and we took the position that we would not add code to support alphas. For consistency with that, I'd be inclined to say that we don't support using the server side dry run against a 1.12 cluster, even if it is enabled. 

> But you remind me, we can do a version check before. If version is after v1.13 and we find cluster admin disable it, we should catch feature disable error and go back to client dry-run.

I agree with that. So:
- If server version >= 1.13, attempt to use server-side dry run _instead of_ client-side
- If server version < 1.13, or server-side dry run returns the "not supported" error, run client-side dry run

Reply:
I didn't know about our alpha restrictions, and its not in the docs so we should add it to contributing.md.

Reply:
Sounds good, thanks for the reminding and making the decision. I added the version check, we'll run server-dry-run when version >=1.13